nomad/server.go

```diff
@@ -1201,28 +1201,52 @@ func (s *Server) setupRpcServer(server *rpc.Server, ctx *RPCContext) error {
 	// The endpoints are client RPCs and don't include a connection
 	// context. They also need to be registered as streaming endpoints in their
 	// register() methods.
+	//
+	// Streaming RPCs should only be registered once, so before we register
+	// them, we make a check if they haven't been already registered.
 
 	clientAllocs := NewClientAllocationsEndpoint(s)
-	clientAllocs.register()
-	_ = server.Register(clientAllocs)
-
 	fsEndpoint := NewFileSystemEndpoint(s)
-	fsEndpoint.register()
-	_ = server.Register(fsEndpoint)
-
 	agentEndpoint := NewAgentEndpoint(s)
-	agentEndpoint.register()
-	_ = server.Register(agentEndpoint)
 
 	// Event is a streaming-only endpoint so we don't want to register it as a
 	// normal RPC
 	eventEndpoint := NewEventEndpoint(s)
-	eventEndpoint.register()
 
 	// Operator takes a RPC context but also has a streaming RPC that needs to
 	// be registered
 	operatorEndpoint := NewOperatorEndpoint(s, ctx)
-	operatorEndpoint.register()
+
+	// streamingEndpoint is a helper interface to make working with endpoint
+	// structs easier
+	type streamingEndpoint interface {
+		register()
+	}
+
+	streaming_endpoints := map[string]streamingEndpoint{
```

Comment:
Hm, this ends up having to allocate this map every time we connect. Trying to test this without allocation seems like we'd need a lock or an atomic, which would be nice to avoid if we can.

What if we move the streaming registration out of this function entirely? This `setupRpcServer` gets called by both `setupRPC` (which is called once) and for each connection. That is, we can move the streaming endpoint registration into `setupRPC` (or into a new function it calls) and we know it'll only ever get called once.

Reply:
I think I understand better now, thanks. Refactored.